### PR TITLE
fix: use predicate for curve operations

### LINKED ISSUE #5045: Point Addition in Unentered else Block Causes Circuit Failure
### Aim

A point addition inside an else that is **NOT** "entered" will cause the circuit to fail entirely. The simulator seems to apply the checks when generating the witness and if using 0 as the base value just fails. 

I encountered this when having a computation that is only to happen when we have non-zero values, but anyway fails even if unused. A minimal test to reproduce is here. 

```
#[test]
fn weird_add() {
    let a = EmbeddedCurvePoint {
        x: 0x1d8eb4378a3bde41e0b6a9a8dcbd21b7ff9c51bdd6ca13ce989abbbf90df3666,
        y: 0x06075b63354f2504f9cddba0b94ed0cef35fc88615e69ec1f853b51eb79a24a0
    };

    let aa = a.double();

    if aa.x == a.add(a).x {
        println(f"We are not trying to do a double of zero");
    } else {
        println("We are trying to do a double of zero");
        let d = EmbeddedCurvePoint { x: 0, y: 0 }.double();
        println(f"d*2 {d}");
    }
}
```

Similar failure is possible with the `multi_scalar_mul` and `fixed_base_scalar_mul` if providing fields with values > 128 bits inside the if. 

### Expected Behavior

I expected to see 
```bash
We are not trying to do a double of zero
```
And have the test pass. 

### Bug

Instead the test fails with the following log.

```bash
We are not trying to do a double of zero
[noir_aztec] Testing encrypted_logs::payload::weird_add... FAIL
Failed to solve program: 'Failed to solve blackbox function: embedded_curve_add, reason: Point (0000000000000000000000000000000000000000000000000000000000000000, 0000000000000000000000000000000000000000000000000000000000000000) is not on curve'
```

### To Reproduce

1. Run the test
2. Profit???
3.
4.

### Project Impact

Blocker

### Impact Context

We need to use the computation to constrain log production in Aztec.

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Installation Method

None

### Nargo Version

_No response_

### NoirJS Version

_No response_

### Would you like to submit a PR for this Issue?

None

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #5045

## Summary\*
use predicate to set is_infinite to true when side_effects are disabled, to ensure that the curve operation will not fail.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by TomAFrench** [comment]
Can you move this predicate into `flatten_cfg` similarly to how we handle `ToRadix`? We can then add the predicate in SSA rather than ACIRgen

**Comment by guipublic** [comment]
> We should also have a test for these EC operations inside of an inactive block.

Yes, I created one but forgot to commit it. It's done now.

**Comment by guipublic** [comment]
> Can you move this predicate into `flatten_cfg` similarly to how we handle `ToRadix`? We can then add the predicate in SSA rather than ACIRgen.

It's done

File: test_programs/execution_success/regression_5045/src/main.nr [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    if aa.x != a.add(a).x {
+        let bad = EmbeddedCurvePoint { x: 0, y: 5, is_infinite: false };
+        let d = bad.double();
+        let e = dep::std::embedded_curve_ops::multi_scalar_mul(
+            [a, bad],
+            [EmbeddedCurveScalar { lo: 1, hi: 0 }, EmbeddedCurveScalar { lo: 1, hi: 0 }]
+        );
+        assert(e[0] != d.x);
+    }
+}
```
[the remark this answers is not in the document]
Comment (guipublic) [reply test_programs/execution_success/regression_5045/src/main.nr]: indeed,
done

File: compiler/noirc_evaluator/src/ssa/ir/instruction.rs [hunk 2]
```diff
@@ -339,6 +344,9 @@ impl Instruction {
 
             // Some `Intrinsic`s have side effects so we must check what kind of `Call` this is.
             Call { func, .. } => match dfg[*func] {
+                // Explicitly allows removal of unused ec operations, even if they can fail
+                Value::Intrinsic(Intrinsic::BlackBox(BlackBoxFunc::MultiScalarMul))
+                | Value::Intrinsic(Intrinsic::BlackBox(BlackBoxFunc::EmbeddedCurveAdd)) => true,
```
[the remark this answers is not in the document]
Comment (guipublic) [reply compiler/noirc_evaluator/src/ssa/ir/instruction.rs:350]: Some test fails without it (compile_success_empty/intrinsic_die).
It makes sense to me to simplify dead EC operations, but I don't have a strong opinion.

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**